solutions/Z3.Linq/Theorem{T}.cs

```diff
@@ -57,4 +59,18 @@ public Theorem<T> Where(Expression<Func<T, bool>> constraint)
     {
         return new Theorem<T>(base.Context, base.Constraints.Concat(new List<LambdaExpression> { constraint }));
     }
+
+    /// <summary>
+    /// OrderBy query operator, used to optimize a solution using query expression syntax.
+    /// </summary>
+    /// <param name="lambda">Expression representing the value to minimize.</param>
+    /// <returns>Environment type instance with properties set to theorem-satisfying values.</returns>
+    public T OrderBy<TResult>(Expression<Func<T, TResult>> lambda) => Optimize(Optimization.Minimize, lambda);
```

Comment:
The thing to consider here is whether this should return some object that has a `Solve` method on it, such that the experience of constraints versus optimization does not oddly differ. Right now, when using `orderby` it will trigger eager evaluation by calling `Optimize` underneath. A lazy evaluation approach seems more appropriate.

One option is to introduce `ISolvable<T>` with a single `T Solve()` method on it and have `OrderBy[Descending]` return an object implementing that (really just akin to a `Lazy<T>` that triggers a call to `Optimize` when `Solve` is called). `Theorem<T>` would also implement this interface.

Reply:
I've pushed an update that implements what I think you had in mind.